_posts/2017-06-05-rolling-upgrade-part-2.md

```diff
@@ -0,0 +1,119 @@
+- Pha (2): Expand database
+
+Lúc này database sẽ có "new" table và trigger. Nếu trên A và B tạo ra các bản ghi ở "old" table thì trigger sẽ làm nhiệm vụ thay đổi nội dung của bản ghi đó cho phù hợp với "new" table.
+
+Hình ngũ giác màu da cam thể hiện các dữ liệu được tạo ra tại pha expand, và hình tam giác màu da cam thể hiện các dữ liệu được tạo bởi trigger.
+
+Hình ngũ giác màu xanh da trời thể hiện các dữ liệu đã có sẵn trước pha expand.
+
+- Pha (3): Migrate database
+
+Lấy toàn bộ dữ liệu ở "old" table trước thời điểm tạo trigger sau đó thay đổi thành dữ liệu phù hợp với "new" table. Nghĩa là tất cả các hình ngũ giác màu xanh da trời sẽ được ánh xạ thành hình tam giác tương ứng tại "new" table.
+
+- Pha (4): Deploy
+
+Nhờ có tính năng rolling upgrade mà ta không phải upgrade hai service cùng một thời điểm mà ta có thể làm lần lượt:
+
+*Thời điểm 1:* Upgrade service A lên phiên bản (N+1):
+A sẽ đọc ghi vào "new" table, và khi tạo ra một bản ghi mới ở "new" table thì cũng sẽ được trigger thay đổi phù hợp rồi ghi vào "old" table.
+Đối với B thì vẫn đọc tại "old" table và nếu ghi vào "old" table thì cũng được trigger thực hiện tương tự sang "new" table.
+
+*Thời điểm 2:* Upgrade service B lên phiên bản (N+1):
+Lúc này tất cả các serivice đã được upgrade lên (N+1)
+
+- Pha (5): Contract database
+
+Sau khi đã upgrade toàn bộ service lên (N+1) thì "old" table sẽ không được đọc ghi bởi bất kì service nào nữa, tất cả đã chuyển sang đọc ghi "new" table
+Thực hiện contract phase sẽ xóa trigger và xóa "old" table
+
+- Pha (6): Hoàn thành
+
+Chúng ta đã hoàn thành quá trình upgrade hệ thống sử dụng trigger để xử lý vấn đề thay đổi database. Trong toàn bộ quá trình, service tại phiên bản N hay (N+1) cũng đều có thể tương tác với database, do vậy sẽ giảm được downtime hệ thống khi upgrade. Ngoài ra, chúng ta cũng có một hướng tiếp cận khác như triggerless, nó cũng được áp dụng cho GitHub, Facebook [2].
+
+Tiếp theo tôi sẽ phân tích tính năng thứ hai.
+
+### 2. Maintenance Mode
+Là chế độ đặt một service vào trạng thái duy trì, bảo hành sửa chữa để cho hệ thống không đặt resources trên node đó nữa khi có yêu cầu từ người dùng tạo mới resources. Cụ thể như Nova trong OpenStack đã áp dụng tính năng này.
+
+Trong Nova có các thành phần như: nova-api, nova-conductor, nova-scheduler, nova-compute,...
+
+Trong đó:
+- nova-compute dùng để tạo ra các máy ảo, vậy máy ảo ở đây chính là resource.
+- nova-scheduler sẽ làm nhiệm vụ tính toán để đặt con máy ảo trên nova-compute nào cho phù hợp.
+
+Vậy câu chuyện là khi người vận hành muốn bảo trì một node nova-compute 30 thì đầu tiên họ phải ra lệnh cho nova kích hoạt chế độ **maintenance mode** cho node nova-compute 30 lên, lúc đó nova-scheduler sẽ không đặt việc tạo máy ảo trên node nova-compute 30 nữa kể cả tài nguyên trên node đó vẫn còn rất nhiều, khi có yêu cầu từ người dùng tạo máy ảo. Để người vận hành có thể thực hiện bảo trì với node nova-compute 30.
+Đó chính là ý nghĩa của tính năng **Maintenance Mode**.
```

Comment:
@NguyenHoaiNam sau khi dẫn dắt loằng ngoằng ở trên, em chốt cho 1 câu `Để abc xyz`. Anh đọc thì sẽ hiểu là `Để abcxyz thì ta cần...` mà hoàn toàn ko thấy đoạn sau ở đâu.
Đây chỉ là cách hành văn cơ bản. Tại sao ko sửa thành 
`Vậy câu chuyện là khi người vận hành muốn bảo trì một node nova-compute 30 thì đầu tiên họ phải ra lệnh cho nova kích hoạt chế độ **maintenance mode** cho node nova-compute 30 lên, lúc đó nova-scheduler sẽ không đặt việc tạo máy ảo trên node nova-compute 30 nữa kể cả tài nguyên trên node đó vẫn còn rất nhiều, khi có yêu cầu từ người dùng tạo máy ảo để người vận hành có thể thực hiện bảo trì với node nova-compute 30.`